### ISSUE TITLE
New Targets creation

### ISSUE THREAD

**Issue by savroff** [stimulusjs/stimulus#41]
This time I will try to act smarter and ask before write code 😂

Sometimes I need to create a DOM elements, that Stimulus controller should controller after.
For example in this interface, when I create a new item, I need a way to add it to DOM.
![vr9svsvjo7](https://user-images.githubusercontent.com/1442465/34625694-b97b9ed8-f227-11e7-9787-4eba02e0e179.gif)

Right now I need to write this

```
    // create tag element
    const item = document.createElement('div')
    item.setAttribute('data-title', 'aa')
    item.setAttribute('data-id', 123)
    item.setAttribute('data-target', 'categories.selected_item')
    item.setAttribute('data-action', 'click->categories#removeItem')
    item.setAttribute('class', 'categories--selected-item')
    item.innerHTML = 'aa'

    this.targets.find('selected_items').appendChild(item)
```

Can we present some sort of function to simply create new dom elements inside. Like:

```
// element can be a 'div' by default
const target = this.targets.create('selected_item', { 
  element: 'div', 
  data: {}, 
  htmlOptions: { class: 'categories--selected-item' }, 
  innerHTML: '' 
})
this.targets.find('selected_items').appendChild(target)
```
  
----------------------------------------------

Probably we can just create a simple standard way to adding mix-ins to controllers.
This case we can keep things like that outside of Stimulus

**Comment by mdesantis** [comment]
What about using `template` HTML5 tag? https://www.html5rocks.com/en/tutorials/webcomponents/template/

**Comment by savroff** [comment]
@mdesantis thank you! 🎉

**Comment by javan** [comment]
You can access a controller's own identifier with `this.identifier`. So you can avoid hard coding and repeating it by doing:
```js
item.setAttribute('data-target', `${this.identifier}.selected_item`)
item.setAttribute('data-action', `click->${this.identifier}#removeItem`)
```
Instead of:
```js
 item.setAttribute('data-target', 'categories.selected_item')
 item.setAttribute('data-action', 'click->categories#removeItem')
```

**Comment by savroff** [comment]
@javan what I actually did.
I created own `Conroller`class with wrapping original Stimulus one. Can we have think about simple way to do a mix-in? Like for example in [Vue](https://vuejs.org/v2/guide/mixins.html)

**Comment by savroff** [comment]
@mdesantis one problem with templates, seems like they not supported in IE 💩 
But a lot of our users still use last version of IE

**Comment by mdesantis** [comment]
Yes, but there are many working polyfills in the wild :-)

**Comment by mdesantis** [comment]
Anyway, I feel the needing of managing potentially existing HTML components, in a library like this one. Something that resolves the requirement "How to display a new incoming message on my instant messaging app?"

**Comment by savroff** [comment]
Agree, this can be placed in examples and in the handbook. 
I don't think that we need something too complex. 
@mdesantis idea with templates is really nice, this case you can keep HTML related things in Rails view, but we need to include polyfill in a framework (could increase library size a little).

Or we can create new elements and insert them like I showed, plus wrap it in nicer helper. But  I see a downside here, that new element markup fully created in the `.js` file.

@javan what you think about all these?

**Comment by javan** [comment]
Remember:
> Stimulus is a JavaScript framework with modest ambitions. It doesn't seek to take over your entire front-end—in fact, it's not concerned with rendering HTML at all.

We don't plan to add our own way to do rendering, and there are a number of great options for you to chose from. I'd start small with `document.createElement()` or `element.cloneNode()`, and then reach for a templating library if needed. I'm a fan of [hyperHTML](https://github.com/WebReflection/hyperHTML) and good ol' [EJS](https://github.com/tj/ejs).

**Comment by pacMakaveli** [comment]
@savroff on top of using template, you can also combine the template functionality with lodash's template and create this: 

javascript: https://github.com/games-directory/games.directory/blob/v2.0-beta1/app/javascript/packs/controllers/identity_verification_process_controller.js

view: https://github.com/games-directory/games.directory/blob/v2.0-beta1/app/views/users/form/verification/_search.erb

The only downside is, at least for me, that you have to use .erb.

TL;DR

You can create a simple HTML template enclosed by a script tag which gets ignored when viewing the page because of `type='text/template'`

```html
<div id='responseBody' class='row align-top'></div>

<script type='text/template' id='responseBodyTemplate'>

  <div class='cell-sm-12 -margin-top_2'>
    
    {{ _.forEach(identities, (identity) => { }}

      <h3 class='-line-reset'>{{= identity.name }}</h3>
      <p class='-line-reset -margin-top_4'>
        About: 
        <span class='-weight-regular'>{{= identity.about }}</span>
      </p>

    {{ }) }}

  </div>  <!-- cell-sm-12 -->

</script>
```

``` js
// Find the Template
const container = document.getElementById('responseBodyTemplate');

// Tell lodash to grab the contents of the Template
const template  = _.template(container.innerHTML);

// Magic happens
// 'parse' the template and pass the required variables
document.getElementById('responseBody').innerHTML = template({
  identities: [{ name: 'Foo', about: 'Bar' }]
});
```

**Comment by savroff** [comment]
@pacMakaveli thanks! 
I personally like keep markup outside from js files. 😍  We similar approach to [this one](https://evilmartians.com/chronicles/evil-front-part-1) and keep HTML in erb sounds really reasanoble in my case.

**Comment by rainerborene** [comment]
Just for posterity I have found two libraries that does exactly what I was looking for: [bind.js](https://github.com/remy/bind.js) and [simulacra](https://github.com/gr0uch/simulacra). One of the authors published an article explaining how it works: [Tiny two way data binding](https://remysharp.com/2015/06/02/bind). It would be very simple to integrate with a stimulus controller and `template` tags.
